GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs

```diff
@@ -0,0 +1,466 @@
+            ITracer tracer,
+            NugetUpgraderConfig config,
+            bool dryRun,
+            PhysicalFileSystem fileSystem,
+            NuGetFeed nuGetFeed,
+            LocalUpgraderServices localUpgraderServices)
+        {
+            this.dryRun = dryRun;
+            this.nugetUpgraderConfig = config;
+            this.tracer = tracer;
+            this.installedVersion = new Version(currentVersion);
+
+            this.fileSystem = fileSystem;
+            this.nuGetFeed = nuGetFeed;
+            this.localUpgradeServices = localUpgraderServices;
+        }
+
+        public string DownloadedPackagePath { get; private set; }
+
+        public static NuGetUpgrader Create(
+            ITracer tracer,
+            bool dryRun,
+            bool noVerify,
+            out string error)
+        {
+            NugetUpgraderConfig upgraderConfig = new NugetUpgraderConfig(tracer, new LocalGVFSConfig());
+            bool isConfigured;
+            bool isEnabled;
+
+            if (!upgraderConfig.TryLoad(out isEnabled, out isConfigured, out error))
+            {
+                if (isEnabled && !isConfigured)
+                {
+                    tracer.RelatedWarning($"NuGetUpgrader is enabled, but is not properly configured. Error: {error}");
+
+                    return new NuGetUpgrader(
+                        ProcessHelper.GetCurrentProcessVersion(),
+                        tracer,
+                        upgraderConfig,
+                        dryRun,
+                        ProductUpgraderInfo.GetAssetDownloadsPath(),
+                        personalAccessToken: null);
+                }
+
+                return null;
+            }
+
+            if (!TryGetPersonalAccessToken(
+                GitBinPath,
+                upgraderConfig.FeedUrlForCredentials,
+                tracer,
+                out string token,
+                out error))
+            {
+                tracer.RelatedWarning($"NuGetUpgrader was not able to acquire Personal Access Token to access NuGet feed. Error: {error}");
+            }
+
+            NuGetUpgrader upgrader = new NuGetUpgrader(
+                ProcessHelper.GetCurrentProcessVersion(),
+                tracer,
+                upgraderConfig,
+                dryRun,
+                ProductUpgraderInfo.GetAssetDownloadsPath(),
+                token);
+
+            return upgrader;
+        }
+
+        public void Dispose()
+        {
+            this.nuGetFeed?.Dispose();
+        }
+
+        public bool UpgradeAllowed(out string message)
+        {
+            if (string.IsNullOrEmpty(this.nugetUpgraderConfig.FeedUrl))
+            {
+                message = "Nuget Feed URL has not been configured";
+                return false;
+            }
+            else if (string.IsNullOrEmpty(this.nugetUpgraderConfig.PackageFeedName))
+            {
+                message = "URL to lookup credentials has not been configured";
```

Comment:
Is this a copy paste error? 

I noticed this is the same message that we return in the ` else if (string.IsNullOrEmpty(this.nugetUpgraderConfig.FeedUrlForCredentials))`case.